a-star_test.go

```diff
@@ -116,3 +119,240 @@ func ExampleShortestPathWithHeuristic() {
+	if nodePos% columns == 1 {
+		return nodePos + (columns - 1)
+	}
+	return nodePos - 1
+}
+
+func getDown(nodePos, rows, columns int) int {
+	FirstNodeOfLastRow := rows*columns - columns + 1
+	if nodePos >= FirstNodeOfLastRow {
+		return nodePos - (rows*columns - columns)
+	}
+	return nodePos + columns
+}
+
+func BenchmarkShortestPathWithHeuristic_100X100GridOfNodes(b *testing.B) {
+	rows, columns := 100, 100
+	grid := createGrid(rows, columns)
+	randomNodes := createLongListOfRandomKeyNodes(rows * columns, 100000)
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		_, ok := grid.ShortestPathWithHeuristic(randomNodes[i].fromNode, randomNodes[i].toNode, heuristicFor100X100Grid)
+		if !ok {
+			b.Fatal(`ok is false`, randomNodes[i])
+		}
+	}
+}
+
+type randFromTo struct {
+	fromNode, toNode string
+}
+
+func createLongListOfRandomKeyNodes(totalNumberOfNodes, resultLength int) (res []randFromTo) {
+	rand.Seed(time.Now().UnixNano())
+
+	for i:=0; i < resultLength; i++ {
+		randRes := randFromTo{"0", "0"}
+		for ; randRes.fromNode == "0" || randRes.toNode == "0" ; {
+			randRes = randFromTo{
+				fromNode: strconv.Itoa(rand.Intn(totalNumberOfNodes)),
+				toNode:   strconv.Itoa(rand.Intn(totalNumberOfNodes)),
+			}
+		}
+
+		res = append(res, randRes)
+	}
+
+	return res
+}
+
+func TestHeuristicCalculationFor100X100Grid(t *testing.T) {
+	expected := []struct {
+		startNode, endNode string
+		distance int
+	}{
+		{
+			startNode: "0",
+			endNode:   "1",
+			distance:  1,
+		},
+		{
+			startNode: "1",
+			endNode:   "3",
+			distance:  2,
+		},
+		{
+			startNode: "1",
+			endNode:   "101",
+			distance:  1,
+		},
+		{
+			startNode: "220",
+			endNode:   "450",
+			distance:  30,
+		},
+		{
+			startNode: "50",
+			endNode:   "850",
+			distance:  8,
+		},
+		{
+			startNode: "950",
+			endNode:   "102",
+			distance:  48,
+		},
+		{
+			startNode: "940",
+			endNode:   "7172",
+			distance:  38, // 32 to the right + 62 down (38 up)
+		},
+		{
+			startNode: "7172",
+			endNode:   "940",
+			distance:  38, // 32 to the left + 62 up (38 down)
+		},
+		{
+			startNode: "7178",
+			endNode:   "7920",
+			distance:  42, // 58 to the left >
+		},
+	}
+
+	for _, e := range expected {
+		res := heuristicFor100X100Grid(e.startNode, e.endNode)
+		if res != e.distance {
+			t.Fatalf("Distance between startNode %v and endNode %v should have been %v but was %v",
+				e.startNode, e.endNode, e.distance, res)
+		}
+	}
+}
+
+func heuristicFor100X100Grid(startKey, endKey string) int {
+	rows, cols := 100, 100
+
+	startPos, _ := strconv.Atoi(startKey)
+	endPos, _ := strconv.Atoi(endKey)
+
+	rowsDiff := calculateDiff((startPos / rows) - (endPos / rows), cols)
+	colsDiff := calculateDiff((startPos % cols) - (endPos % cols), cols)
+
+	var distanceToEndNode int
+	for ; rowsDiff > 0 || colsDiff > 0; {
+		// both diff's get reduced by one to simulate a diagonal step
+		rowsDiff--
+		colsDiff--
+
+		distanceToEndNode++
+	}
+
+	return distanceToEndNode
```

Comment:
The for declaration which end at line 345 is going on like this:
`for ; rowsDiff > 0 || colsDiff > 0; {`
so when it ends both rowsDiff and colsDiff r 0 or negative (which for the function needs is 0) so it's seems to be fine as is

Reply:
Actually when looking at it with fresh eyes it can just be without the for and instead:
`
if rowsDiff > colsDiff {
  return rowsDiff
}
return colsDiff
`

The former version is the kind of code I get when I'm tired...

Reply:
You're right, the original loop is fine, I misread the loop condition. I thought it stopped when either of the diffs reached 0.

` if rowsDiff > colsDiff { return rowsDiff } return colsDiff` is wrong though, as you would only walk one side of an L-shaped path, and that the distance can't be right, since the diagonal is always the longest side in a triangle like this.

To simplify this code, we could do:
```
distanceToEndNode := 0
if rowsDiff > colsDiff {
    distanceToEndNode = colsDiff + (rowsDiff - colsDiff)
} else {
    distanceToEndNode = rowsDiff + (colsDiff - rowsDiff)
}
```

That way, we use the shorter dimension of the L to know how many diagonal steps we take, and then walk along the remaining part of the longer side of the L.

Reply:
As far as I understand ur code (and I do need coffee so maybe I don't) it gets the same result as mine, it's going something along this line:
```
if x > y {
    distance = y + (x - y) // which is x - y + y so distance equal x
} else {
    distance = x + (y - x) // which is y - x + x so distance equal y
}
```

In regard to the comment that diagonal side is longest side in a triangle, well I won't disagree that ;) but my thoughts were less in regard real distance imitation but more in the direction of a grid which allow a "unit" to move to the sides and also diagonally in one move

Reply:
Oh man, you're completely right of course... 👍